### ISSUE TITLE
ln: failed to create symbolic link '/var/run/zrep.lock/27675': No such file or directory

### ISSUE THREAD

**Issue by ams-tschoening** [bolthole/zrep#92]
Once a while I get the following error during execution of `zrep` using `cron`:

    ln: failed to create symbolic link '/var/run/zrep.lock/27675': No such file or directory

My `crontab` looks like the following:

	6,16,26,36,46,56 3-23 * * * /usr/sbin/zrep -S zfs-pool/vbox/kid >/dev/null
	 1 0,3-23 * * * /usr/bin/backup_to_zfs-pool_system_kid >/dev/null
	11 0,3-23 * * * /usr/sbin/zrep -S zfs-pool/system/kid  >/dev/null

I already had a look at [similar issues](https://github.com/bolthole/zrep/issues/8) and the [code handling the global lock](https://github.com/bolthole/zrep/blob/master/zrep#L287), but have some difficulties in understanding how things are supposed to work:

Is `zrep.lock` supposed to be a directory or a file?

Because the error message speaks of a directory, but that doesn't exist in my case by default and I don't see any manual creation of that in the code. Additionally, the lock handling uses things like `rm` which doesn't delete directories by default, so I would expect things to fail there. OTOH, the used `ln -s` would work if `zrep.lock` was a directory and simply create files in that. But again, `rm ...` in [releasing a lock](https://github.com/bolthole/zrep/blob/master/zrep#L318) is not releasing the symbolic links in a directory, but removing the directory itself:

    rm $Z_GLOBAL_LOCKFILE

I don't see any directory name+file concatenation on that variable as well. I guess the error message is simply STDERR of the [following line](https://github.com/bolthole/zrep/blob/master/zrep#L287):

    $LN_S /proc/$Z_GLOBAL_PID $Z_GLOBAL_LOCKFILE && return 0

So it might be that things work like expected and I only see error messages which could easily be ignored? In which case you might want to consider redirecting them to `/dev/null`?

Thanks for any help!

**Comment by tschoening** [comment]
> Are you running the most recent code in github?

No, I'm using 1.7.3 and looking at the commit history I guess you have the following in mind?

> added force of ln -n, to make lock work properly on all systems

https://github.com/bolthole/zrep/commit/d2acc8cc6f5f7c82d6ec9626873b9bc4e6b6d676

In my case `zrep.lock` is no link to a directory, it's missing by default. So you create that file as a link as part of locking, which is wrongly assumed to be some directory on additional invocations of `ln -s` without using `-n`? So, should I simply try to run `master` then and see if it fixes things?

OTOH, is that a sign of a wrong setup already or supported, because that's exactly what the lock is about? It detects concurrent invocations of `zrep` and handles those, only with unnecessary error messages because of details of `ln -s`? This problem most likely only happens once per day.

**Comment by ppbrown** [comment]
well, the error messages arent unneccessary. they are correctly telling you
something is wrong :)
yes the latest should fix your issue I think.


On Mon, Jun 18, 2018 at 7:04 AM, Thorsten Schöning <notifications@github.com

**Comment by tschoening** [comment]
After upgrading to current master, the error message changed but the problem still persists:

    ln: failed to create symbolic link '/var/run/zrep.lock': File exists

Looking at the [following line](https://github.com/bolthole/zrep/blob/master/zrep#L287), that message makes sense to me, because I don't see where the lock is checked if it exists already before trying to acquire it.

    $LN_S /proc/$Z_GLOBAL_PID $Z_GLOBAL_LOCKFILE && return 0

While there is the function [zrep_has_global_lock](https://github.com/bolthole/zrep/blob/master/zrep#L262), it doesn't seem to be used wherever [zrep_get_global_lock](https://github.com/bolthole/zrep/blob/master/zrep#L280) is used, so a still existing lock during acquiring one seems to be a valid use case to me. Shouldn't the above line simply redirect its error message, especially because the code [afterwards](https://github.com/bolthole/zrep/blob/master/zrep#L299) handles existing locks using a loop?

    while (( retry_count > 0 )); do

You redirect errors in other commands as well, e.g. in [zrep_global_lock_pid](https://github.com/bolthole/zrep/blob/master/zrep#L256):

    $LS -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'

**Comment by ppbrown** [comment]
Hmm.
you say "the problem still persists".

I think a more accurate statement would be "there is still a problem".
However, at least it is not the SAME problem :)
Well, potentially two problems, from your point of view:

1. you dont like the error message from ln, i'm guessing. So you view that
as a problem?
   However, I view that as valuable debugging information ;)
  At least the first time. While doing the retry loop, probably not so
useful.

2. you're running into lock contention

My point of view on the error is that if the user is doing things "right"..
they should never encounter the lock contention.
So, for at least the first time, that error gives the user a hint of where
to look, for what is going on, without having to wade through lots of zrep
documentation.


Comments?
Do you think there is still some kind of bug in the code, or do you think
the lock contention error is valid?

**Comment by tschoening** [comment]
That depends on if `zrep`  is allowed to run more than once and should handle that gracefully itself. If so, in my opinion it should not print error messages for `ln` during acquiring the lock, because that this might fail is expected behaviour. You have other error messages in place if it fails too often and such and those are fine for me.

If it shouldn't ever run more than once, I don't understand why there's a loop for the lock anyway, it could just fail fast instead, because no one else should hold the lock.

Additionally, the posted error happens only once for me always, so from my understanding it's using the loop afterwards and one second later the lock is available, so all former instances of `zrep` have finished their work properly. I don't get any other error messages.

If you have a look at my initially posted `crontab`, you can see that it's perfectly possible that multiple instances overlap for some time, so you need to decide if I'm using `zrep` totally wrong or not. If not, as said, I would vote for getting rid of the error messages of `ln`, because they are expected and nothing to worry about. You handle that case properly then.

**Comment by ppbrown** [comment]
Thank you for the diagnosis.
I think that it is important to know when zrep is overlapping to the point of contention. Most people would want to know, and fine-tune their scheduling, so I will choose to keep the error messages in the main code.

**Comment by ams-tschoening** [comment]
Just for my personal bookkeeping, the following PR might be of interest as well:

https://github.com/bolthole/zrep/pull/77

**Comment by ppbrown** [comment]
what would be even more of interest, would be the latest release, which doesnt use ln or /proc ;)

v1.7.5  https://github.com/bolthole/zrep/commit/63e83888b295e46d0a71dbc6cc1a97b6f6f95c84

**Comment by ams-tschoening** [comment]
The latest release doesn't change much, there's still an in my opinion unnecessary error message regarding the global lock:

> /usr/local/zrep/zrep[320]: /var/run/zrep.lock: file already exists [File exists]

    set -C
    echo $Z_GLOBAL_PID > $Z_GLOBAL_LOCKFILE && return 0

https://github.com/bolthole/zrep/blob/master/zrep#L327

> -C
Prevent output redirection using ‘>’, ‘>&’, and ‘<>’ from overwriting existing files.

https://www.gnu.org/software/bash/manual/bash.html#The-Set-Builtin

**Comment by ppbrown** [comment]
Thank you for the update.
There are three typical reasons why that type of output occurs:

1. there is an incompatibility bug between the code, and the OS, reguarding
lock handling.
2. the user is running zrep so often, that their jobs are not completing.
Depending on intent, that could be considered a serious error
3. zrep is crashing and not cleaning up the lockfile

All of these situations seem to be highly important to me, and justify the
user seeing output in those instances.

As the comments in that section of the code say,

#We do NOT try to clean up stale global lock.
#This is a shortterm lock. It should never be stale. If it is,
#it could indicate a more serious system/zrep problem happening.

So it is troubling that you are seeing this output.

I will note that the error you are seeing, is different from the errors you
saw when you initially opened this issue.
If you would like to pursue it with me (and I encourage you to do so), then
I would suggest opening a new ticket for it.
